File: WDF-Server/src/main/java/com/nohorang/suryongfootprint/service/UserService.java
package com.nohorang.suryongfootprint.service;

import com.nohorang.suryongfootprint.model.User;
import com.nohorang.suryongfootprint.model.request.UserCreationRequest;
import com.nohorang.suryongfootprint.model.request.UserLoginRequest;
import com.nohorang.suryongfootprint.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.val;
import org.springframework.beans.BeanUtils;
import org.springframework.stereotype.Service;

import javax.persistence.EntityExistsException;
import javax.persistence.EntityNotFoundException;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;

    //회원가입
    public User createUser(UserCreationRequest request){
        User user = new User();
        BeanUtils.copyProperties(request, user);
        return userRepository.save(user);
    }

    //로그인
    public User getUserLogin(UserLoginRequest loginRequest){
        Optional<User> user = userRepository.findById(loginRequest.getUserId());
        if(user.isPresent()){
            if(user.get().getUserPw().equals(loginRequest.getUserPw())) {
                return user.get();
            }
        }
        //로그인 실패
        throw new EntityNotFoundException("Cant find any user under given ID and PW");
    }

    //아이디 찾기
    public String findUserId(String user_name,String user_email){
        List<User> user = userRepository.findByUserNameAndUserEmail(user_name,user_email);
        if(user.get(0)!=null){
            return user.get(0).getUserId();
        }
        //아이디 찾기 실패
        throw new EntityNotFoundException("Cant find any user given Info");
    }

    //비밀번호 찾기
    public String findUserPW( String user_name, String user_id,String user_email){
        Optional<User> user = userRepository.findByUserIdAndUserNameAndUserEmail(user_id,user_name,user_email);
        String pw_str = "";
        if(user.isPresent()){
            pw_str = user.get().getUserPw();
            return pw_str.substring(0,2)+"*".repeat(pw_str.length()-2);
        }
        //비밀번호 찾기 실패
        throw new EntityNotFoundException("Cant find any user given Info");
    }

    //비밀번호 변경
    public User updateUserPW(String user_id, UserCreationRequest request){
        Optional<User> user = userRepository.findById(user_id);
        if (!user.isPresent()) {
            throw new EntityNotFoundException("User Not Found");
        }
        User c_user = user.get();
        c_user.setUserPw(request.getUserPw());
        return userRepository.save(c_user);
    }

    //닉네임 변경
    public User updateUserNickName(String user_id, UserCreationRequest request){
        Optional<User> user = userRepository.findById(user_id);
        if (!user.isPresent()) {
            throw new EntityNotFoundException("User Not Found");
        }
        User c_user = user.get();
        c_user.setUserNickname(request.getUserNickname());
        return userRepository.save(c_user);
    }

    //회원 탈퇴
    public void deleteUser(String user_id){
        userRepository.deleteById(user_id);
    }

    //아이디 중복 조회
    public String existUserId(String user_id){
        Optional<User> user = userRepository.findById(user_id);
        if (user.isPresent()) {
            return user.get().getUserId();
        }
        throw new EntityNotFoundException("Cant find any user given Info");
    }

    //이메일 중복 조회
    public String existUserEmail(String user_email){
        Optional<User> user = userRepository.findByUserEmail(user_email);
        if (user.isPresent()) {
            return user.get().getUserEmail();
        }
        throw new EntityNotFoundException("Cant find any user given Info");
    }

    //닉네임 중복 조회
    public String existUserNickName(String user_nickname){
        Optional<User> user = userRepository.findByUserNickname(user_nickname);
        if (user.isPresent()) {
            return user.get().getUserNickname();
        }
        throw new EntityNotFoundException("Cant find any user given Info");
    }

}

File: WDF-Android/app/src/main/java/kr/co/nohorang/suryongfootprint/api/UserService.kt
package kr.co.nohorang.suryongfootprint.api

import kr.co.nohorang.suryongfootprint.data.LoginRequestDTO
import kr.co.nohorang.suryongfootprint.data.User
import retrofit2.Call
import retrofit2.http.*

//사용할 HTTP CRUD 동작들을 정의한 인터페이스
interface UserService {
    //회원가입 - 정상 수행됨
    //request body - 회원가입할 User 정보
    @POST("/s-footprint/user")
    fun createUser(@Body user : User): retrofit2.Call<User>

    //로그인
    //request body - 아이디, 비밀번호
    @POST("/s-footprint/user/login")
    fun doLogin(@Body loginInfo : LoginRequestDTO): retrofit2.Call<User>

    //아이디 찾기
    @GET("/s-footprint/user/user_id")
    fun findUserId(@Query("user_name") user_name : String,
                   @Query("user_email") user_email : String
    ):Call<String>

    //비밀번호 찾기
    @GET("/s-footprint/user/user_pw")
    fun findUserPW(@Query("user_name") user_name : String,
                   @Query("user_id") user_id : String,
                   @Query("user_email") user_email : String
    ):Call<String>

    //닉네임 변경하기
    //Body - User에서 user_nickname값만 사용 -> 나머지는 null 값이어도 됨.
    @PATCH("/s-footprint/user/{user_id}/nickname")
    fun updateNickname(@Path("user_id") user_id:String,
                        @Body user:User) : Call<User>

    //비밀번호 변경하기
    //Body - User에서 user_pw값만 사용 -> 나머지는 null 값이어도 됨.
    @PATCH("/s-footprint/user/{user_id}/pw")
    fun updatePW(@Path("user_id") user_id:String,
                @Body user:User) : Call<User>

    //회원 탈퇴
    @DELETE("/s-footprint/user/{user_id}")
    fun deleteUser(@Path("user_id") user_id:String):Call<Void>

    //아이디 중복 조회
    @GET("/s-footprint/user/available/id")
    fun existUserId(@Query("user_id") user_id:String):Call<String>

    //이메일 중복 조회
    @GET("/s-footprint/user/available/email")
    fun existUserEmail(@Query("user_email") user_email:String):Call<String>

    //닉네임 중복 조회
    @GET("/s-footprint/user/available/nickname")
    fun existUserNickname(@Query("user_nickname") user_nickname:String):Call<String>
}